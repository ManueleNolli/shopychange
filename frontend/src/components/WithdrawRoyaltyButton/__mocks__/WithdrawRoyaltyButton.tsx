import { divWithoutChildrenMock } from '../../../../mocksUtils'

const WithdrawRoyaltyButtonMock = () =>
  divWithoutChildrenMock('WithdrawRoyaltyButton')
export default WithdrawRoyaltyButtonMock
