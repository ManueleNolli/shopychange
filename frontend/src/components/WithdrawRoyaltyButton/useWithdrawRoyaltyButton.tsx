import { useToast } from '@chakra-ui/react'
import { useState, useEffect } from 'react'
import { useUserContext } from '../../context/userContext'
import {
  tokenRoyaltiesQuery,
  collectionRoyaltiesQuery,
  handlerQueryErrorCatch,
} from '../../services/BackendService'
import {
  withdrawRoyalty,
  waitTransaction,
} from '../../services/BlockchainService'
import { WithdrawReceiver } from '../../types/components/WithdrawReceiver'
import { WithdrawRoyaltyButtonProps } from '../../types/props/WithdrawRoyaltyButtonProps'
import useLoading from '../../hooks/useLoading'
import useFetching from '../../hooks/useFetching'

export default function useWithdrawRoyaltyButton({
  collectionAddress,
  tokenId,
}: WithdrawRoyaltyButtonProps) {
  const { userAddress, blockchainNetworkId } = useUserContext()

  const toast = useToast()
  const [withdrawValue, setWithdrawValue] = useState(0)
  const { isLoading, setIsLoading } = useLoading()
  const { isFetching, setIsFetching, isError, setIsError } = useFetching()
  const [withdrawInfo, setWithdrawInfo] = useState({
    paymentSplitterAddress: '',
    index: 0,
  })
  const [hasPaymentSplitter, setHasPaymentSplitter] = useState(false)

  useEffect(() => {
    const fetchReceiver = async () => {
      if (!userAddress || !blockchainNetworkId) return

      let receivers = []
      let paymentSplitterAddress = ''

      if (tokenId !== undefined) {
        const response = await handlerQueryErrorCatch(
          tokenRoyaltiesQuery(blockchainNetworkId, collectionAddress, tokenId)
        )

        if (response.error) {
          setIsFetching(false)
          setIsError(true)
          return
        }

        const result = response.result
        receivers = result.withdrawRoyaltyToken.receivers
        paymentSplitterAddress =
          result.withdrawRoyaltyToken.paymentSplitterAddress
      } else {
        const response = await handlerQueryErrorCatch(
          collectionRoyaltiesQuery(blockchainNetworkId, collectionAddress)
        )
        if (response.error) {
          setIsFetching(false)
          setIsError(true)
          return
        }

        const result = response.result

        receivers = result.withdrawRoyaltyCollection.receivers
        paymentSplitterAddress =
          result.withdrawRoyaltyCollection.paymentSplitterAddress
      }
      if (paymentSplitterAddress === null) {
        return
      }

      const filter = receivers.filter((receiver: WithdrawReceiver) => {
        return receiver.receiver === userAddress
      })

      if (filter.length > 0) {
        const index = receivers.indexOf(filter[0])
        setWithdrawValue(filter[0].amount)
        setWithdrawInfo({
          paymentSplitterAddress: paymentSplitterAddress,
          index: index,
        })
        setHasPaymentSplitter(true)
      }

      setIsFetching(false)
    }
    fetchReceiver()
  }, [userAddress])

  const withdraw = async () => {
    setIsLoading(true)
    withdrawRoyalty(withdrawInfo.paymentSplitterAddress, withdrawInfo.index)
      .then(waitTransaction)
      .then(() => {
        setIsLoading(false)
        setWithdrawValue(0)
        toast({
          variant: 'solid',
          title: 'Transaction successful',
          description: 'Royalties withdrawn successfully',
          status: 'success',
          duration: 9000,
          isClosable: true,
        })
      }) // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .catch((err: any) => {
        // First phrase of the error message is enough
        const errorMessage = err.message.split('.')[0]
        toast({
          variant: 'solid',
          title: 'Transaction failed',
          description: errorMessage,
          status: 'warning',
          duration: 9000,
          isClosable: true,
        })
        setIsLoading(false)
      })
  }

  return {
    isFetching,
    isError,
    withdrawValue,
    hasPaymentSplitter,
    isLoading,
    withdraw,
  }
}
