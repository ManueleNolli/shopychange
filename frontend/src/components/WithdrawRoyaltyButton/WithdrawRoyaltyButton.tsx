import React from 'react'
import { WithdrawRoyaltyButtonProps } from '../../types/props/WithdrawRoyaltyButtonProps'
import RoyaltiesCheck from '../../utils/RoyaltiesCheck/RoyaltiesCheck'

import { colors } from '../../styles/Colors'
import LoadingButton from '../LoadingButton/LoadingButton'
import ShadowButton from '../ShadowButton/ShadowButton'

import useWithdrawRoyaltyButton from './useWithdrawRoyaltyButton'
import QueryContainer from '../../utils/QueryContainer/QueryContainer'

export default function WithdrawRoyaltyButton({
  collectionAddress,
  tokenId,
  buttonProps,
}: WithdrawRoyaltyButtonProps) {
  const {
    isFetching,
    isError,
    withdrawValue,
    isLoading,
    hasPaymentSplitter,
    withdraw,
  } = useWithdrawRoyaltyButton({ collectionAddress, tokenId })

  if (!hasPaymentSplitter) {
    return <></>
  } else {
    return (
      <RoyaltiesCheck
        collectionAddress={collectionAddress}
        tokenId={tokenId}
        redirectHome={false}
      >
        <QueryContainer isLoading={isFetching} isError={isError}>
          <LoadingButton
            isLoading={isLoading}
            loadingButtonProps={
              buttonProps ? { ...buttonProps } : { maxW: '20em', w: '20em' }
            }
          >
            <ShadowButton
              label={`Withdraw ${Number(withdrawValue.toFixed(4))} ETH`}
              bgColor={`${colors.primaryChakraColor}.400`}
              textColor={'white'}
              shadowColor={colors.primaryShadowColor}
              hoverColor={`${colors.primaryChakraColor}.500`}
              focusColor={`${colors.primaryChakraColor}.500`}
              onClick={withdraw}
              buttonProps={
                buttonProps ? { ...buttonProps } : { maxW: '20em', w: '20em' }
              }
            />
          </LoadingButton>
        </QueryContainer>
      </RoyaltiesCheck>
    )
  }
}
