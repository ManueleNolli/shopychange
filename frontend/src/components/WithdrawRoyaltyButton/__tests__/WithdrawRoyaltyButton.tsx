import React from 'react'
import { render } from '@testing-library/react'
import useWithdrawRoyaltyButton from '../useWithdrawRoyaltyButton'
import WithdrawRoyaltyButton from '../WithdrawRoyaltyButton'

jest.mock('../../../utils/RoyaltiesCheck/RoyaltiesCheck', () => ({
  __esModule: true,
  default: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="RoyaltiesCheck"> {children} </div>
  ),
}))

jest.mock('../useWithdrawRoyaltyButton', () => jest.fn())

const mockProps = {
  collectionAddress: '0x123',
  tokenId: 0,
  buttonProps: {
    maxW: '20em',
    w: '20em',
  },
}

describe('WithdrawRoyaltyButton', () => {
  test('renders WithdrawRoyaltyButton component, without payment splitter', () => {
    ;(useWithdrawRoyaltyButton as jest.Mock).mockReturnValue({
      withdrawValue: 0,
      isLoading: false,
      hasPaymentSplitter: false,
      withdraw: jest.fn(),
    })

    const { container } = render(<WithdrawRoyaltyButton {...mockProps} />)
    expect(container).toMatchSnapshot()
  })

  test('renders WithdrawRoyaltyButton component, with payment splitter', () => {
    ;(useWithdrawRoyaltyButton as jest.Mock).mockReturnValue({
      withdrawValue: 0,
      isLoading: false,
      hasPaymentSplitter: true,
      withdraw: jest.fn(),
    })

    const { container } = render(<WithdrawRoyaltyButton {...mockProps} />)
    expect(container).toMatchSnapshot()
  })
})
