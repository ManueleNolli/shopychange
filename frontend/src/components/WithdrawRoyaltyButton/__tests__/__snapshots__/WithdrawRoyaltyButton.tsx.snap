// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithdrawRoyaltyButton renders WithdrawRoyaltyButton component, with payment splitter 1`] = `
<div>
  <div
    data-testid="RoyaltiesCheck"
  >
     
    <button
      class="chakra-button css-7v8t3l"
      data-testid="onClick-button"
      style="white-space: initial;"
      type="button"
    >
      Withdraw 0 ETH
    </button>
     
  </div>
</div>
`;

exports[`WithdrawRoyaltyButton renders WithdrawRoyaltyButton component, without payment splitter 1`] = `<div />`;
